### PR TITLE
fix(sdk): as code fixes and unit tests

### PR DESCRIPTION
**Pull request by fsamin**
1. Description
probably some the bugs has already been fixed in #3494, the most important parts are the unit tests
1. Related issues
1. About tests
1. Mentions

@ovh/cds

### CODE REVIEW

**Comment by sguiheux** [comment]
Please do not merge this PR, i will a to rewrite all the file on my branch

**Comment by fsamin** [comment]
> Please do not merge this PR, i will a to rewrite all the file on my branch

i know their will be a lot of conflicts with your work, that's why i wrote a new test function, so you just have to keep it as it:  i'm pretty sure the bugs i've fixed don't exist on your branch, they were all related with forks and outgoing hooks browsing

File: sdk/exportentities/workflow.go [hunk 1]
```diff
@@ -777,6 +807,33 @@ func (e *NodeEntry) processNodeAncestors(name string, w *sdk.Workflow) (bool, er
...
 }
 
+func (e *NodeEntry) processOutgoingHookAncestor(name string, w *sdk.Workflow) (bool, error) {
+	a := e.DependsOn[0]
+	ancestors := w.OutgoingHooks()
+	var ancestor *sdk.WorkflowNodeOutgoingHook
+
+	for i := range ancestors {
+		if ancestors[i].Name == a {
+			ancestor = &ancestors[i]
```
Comment (bnjjj) [inline sdk/exportentities/workflow.go]: missing break ?

File: sdk/exportentities/workflow.go [hunk 2]
```diff
@@ -680,41 +709,32 @@ func (e *NodeEntry) processOutgoingHook(name string, w *sdk.Workflow) (bool, err
 		WorkflowHookModel: sdk.WorkflowHookModel{Name: e.OutgoingHookModelName},
 	}
 
-	if ancestor.OutgoingHooks == nil {
-		ancestor.OutgoingHooks = []sdk.WorkflowNodeOutgoingHook{}
+	nodeAncestor, ok := ancertor.(*sdk.WorkflowNode)
```
Comment (bnjjj) [inline sdk/exportentities/workflow.go]: s/ancertor/ancestor

File: sdk/exportentities/workflow.go [hunk 3]
```diff
@@ -650,26 +672,33 @@ func (e *NodeEntry) processNode(name string, w *sdk.Workflow) (bool, error) {
...
+	} else if exist {
+		return true, nil
+	}
+
+	return e.processOutgoingHookAncestor(name, w)
 }
 
 func (e *NodeEntry) processOutgoingHook(name string, w *sdk.Workflow) (bool, error) {
-	var ancestor *sdk.WorkflowNode
+	var ancertor interface{}
```
Comment (bnjjj) [inline sdk/exportentities/workflow.go]: s/ancertor/ancestor

File: sdk/exportentities/workflow.go [hunk 4]
```diff
@@ -458,7 +475,11 @@ func (w Workflow) GetWorkflow() (*sdk.Workflow, error) {
 		attempt++
 	}
 	if len(entries) > 0 {
-		return nil, fmt.Errorf("Unable to process %+v", entries)
+		leftEntries := []string{}
```
Comment (bnjjj) [inline sdk/exportentities/workflow.go]: make with capacity to `len(entries)`

File: sdk/exportentities/workflow.go [hunk 5]
```diff
@@ -70,33 +70,58 @@ const (
...
-		ancestors := make([]string, 0, len(ancestorIDs))
-		for _, aID := range ancestorIDs {
-			a := w.GetNode(aID)
-			if a == nil {
-				return entry, sdk.ErrWorkflowNodeNotFound
+	ancestors := []string{}
+
+	//we have to find the ancestors
+	nodes := w.Nodes(true)
+	nodeNames := []string{}
```
Comment (bnjjj) [inline sdk/exportentities/workflow.go]: make with capacity to `len(nodes) - 1`

**Comment by fsamin** [comment]
Guys, according to https://github.com/ovh/cds/pull/3545#issuecomment-435680146 are you waiting me to fix this PR ?
It was very long hours of debugging and implementing (IMO) a great unit tests function. For one week we have bug fixes nearly mergeable. What are we waiting for ?

**Pull request closed without merging**